src/test/groovy/org/prebid/server/functional/tests/DsaSpec.groovy

```diff
@@ -0,0 +1,293 @@
+package org.prebid.server.functional.tests
+
+import org.prebid.server.functional.model.db.StoredRequest
+import org.prebid.server.functional.model.request.amp.AmpRequest
+import org.prebid.server.functional.model.request.auction.BidRequest
+import org.prebid.server.functional.model.request.auction.RegsDsa
+import org.prebid.server.functional.model.request.auction.ReqsDsaRequiredType
+import org.prebid.server.functional.model.response.auction.BidExt
+import org.prebid.server.functional.model.response.auction.BidExtDsa
+import org.prebid.server.functional.model.response.auction.BidResponse
+
+import static org.prebid.server.functional.model.response.auction.ErrorType.GENERIC
+
+class DsaSpec extends BaseSpec {
+
+    def "AMP request should send DSA to bidder and succeed when DSA is not required"() {
+        given: "Default AmpRequest"
+        def ampRequest = AmpRequest.defaultAmpRequest
+        def dsa = RegsDsa.getDefaultRegsDsa(dsaRequired)
+
+        and: "Stored default bid request with DSA"
+        def ampStoredRequest = BidRequest.defaultBidRequest.tap {
+            regs.ext.dsa = dsa
+            setAccountId(ampRequest.account)
+        }
+
+        and: "Save storedRequest into DB"
+        def storedRequest = StoredRequest.getStoredRequest(ampRequest, ampStoredRequest)
+        storedRequestDao.save(storedRequest)
+
+        and: "Default bidder response with DSA"
+        def bidResponse = BidResponse.getDefaultBidResponse(ampStoredRequest).tap {
+            seatbid[0].bid[0].ext = bidExt
+        }
+
+        and: "Set bidder response"
+        bidder.setResponse(ampStoredRequest.id, bidResponse)
+
+        when: "PBS processes amp request"
+        def response = defaultPbsService.sendAmpRequest(ampRequest)
+
+        then: "Bidder request should contain DSA"
+        def bidderRequests = bidder.getBidderRequest(ampStoredRequest.id)
+        verifyAll {
+            bidderRequests.regs.ext.dsa.required == dsa.required
+            bidderRequests.regs.ext.dsa.dataToPub == dsa.dataToPub
+            bidderRequests.regs.ext.dsa.pubRender == dsa.pubRender
+            bidderRequests.regs.ext.dsa.transparency[0].domain == dsa.transparency[0].domain
+            bidderRequests.regs.ext.dsa.transparency[0].params == dsa.transparency[0].params
+        }
+
+        and: "Bidder response should not contain DSA"
+        def bidderResponse = decode(response.ext.debug.httpcalls.get("generic")[0].responseBody, BidResponse)
+        assert bidderResponse.seatbid[0].bid[0].ext?.dsa == null
+
+
```

Comment:
Minor, but there’s a double space. Please use the same approach inside Spec. I suggest changing it to: `assert !bidderResponse.seatbid[0].bid[0].ext?.dsa` Please do the same for other occurrences.